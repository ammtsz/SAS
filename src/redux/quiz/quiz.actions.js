import { QuizActionsTypes } from "./quiz.types";

// REDUCERS
export const actionSetQuizActive = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_ACTIVE,
  payload: data,
});
export const actionSetQuizLoading = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_LOADING,
  payload: data,
});
export const actionSetQuizCategory = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CATEGORY,
  payload: data,
});
export const actionSetQuizDifficulty = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_DIFFICULTY,
  payload: data,
});
export const actionSetQuizQuestionNumber = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUESTION_NUMBER,
  payload: data,
});
export const actionSetQuizPromotion = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_PROMOTION,
  payload: data,
});

export const actionSetQuizCurrentQuestion = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CURRENT_QUESTION,
  payload: data,
});
export const actionSetQuizCurrentOptions = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_CURRENT_OPTIONS,
  payload: data,
});
export const actionSetQuizToken = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_TOKEN,
  payload: data,
});
export const actionSetQuizQuestionsDatas = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUESTIONS_DATAS,
  payload: data,
});
export const actionSetQuizError = (data) => ({
  type: QuizActionsTypes.REDUCER_SET_QUIZ_ERROR,
  payload: data,
});

export const actionResetQuiz = () => ({
  type: QuizActionsTypes.REDUCER_RESET_QUIZ,
});


// SAGAS
export const actionGetNewQuestion = (data) => ({
  type: QuizActionsTypes.SAGA_GET_NEW_QUESTION,
  payload: data,
});
export const actionCheckAnswer = (data) => ({
  type: QuizActionsTypes.SAGA_CHECK_ANSWER,
  payload: data,
});
export const actionUpdateQuizReport = () => ({
  type: QuizActionsTypes.SAGA_UPDATE_QUIZ_REPORT,
});
export const actionUpdateQuizResumeReport = () => ({
  type: QuizActionsTypes.SAGA_UPDATE_QUIZ_RESUME_REPORT,
});
export const actionGoToNextQuestion = (data) => ({
  type: QuizActionsTypes.SAGA_GO_TO_NEXT_QUESTION,
  payload: data,
});
export const actionFinishQuiz = (data) => ({
  type: QuizActionsTypes.SAGA_FINISH_QUIZ,
  payload: data,
});
export const actionResumeQuiz = (data) => ({
  type: QuizActionsTypes.SAGA_RESUME_QUIZ,
  payload: data,
});
