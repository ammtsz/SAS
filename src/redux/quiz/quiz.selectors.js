import { createSelector } from "reselect";

const selectQuiz = (state) => state.quiz;

export const selectQuizActive = createSelector(
  [selectQuiz],
  (quiz) => quiz.active
);
export const selectQuizLoading = createSelector(
  [selectQuiz],
  (quiz) => quiz.loading
);
export const selectQuizCategory = createSelector(
  [selectQuiz],
  (quiz) => quiz.category
);
export const selectQuizDifficulty = createSelector(
  [selectQuiz],
  (quiz) => quiz.difficulty
);
export const selectQuizQuestionNumber = createSelector(
  [selectQuiz],
  (quiz) => quiz.questionNumber
);
export const selectQuizPromotion = createSelector(
  [selectQuiz],
  (quiz) => quiz.promotion
);
export const selectQuizCurrentQuestion = createSelector(
  [selectQuiz],
  (quiz) => quiz.currentQuestion
);
export const selectQuizCurrentOptions = createSelector(
  [selectQuiz],
  (quiz) => quiz.currentOptions
);
export const selectQuizToken = createSelector(
  [selectQuiz],
  (quiz) => quiz.token
);
export const selectQuizQuestionsDatas = createSelector(
  [selectQuiz],
  (quiz) => quiz.questionsDatas
);
