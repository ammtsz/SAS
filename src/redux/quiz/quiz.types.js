export const QuizActionsTypes = {

    REDUCER_SET_ACTIVE: "REDUCER_SET_ACTIVE",
    REDUCER_SET_LOADING: "REDUCER_SET_LOADING",
    REDUCER_SET_CATEGORY: "REDUCER_SET_CATEGORY",
    REDUCER_SET_DIFFICULTY: "REDUCER_SET_DIFFICULTY",
    REDUCER_SET_QUESTION_NUMBER: "REDUCER_SET_QUESTION_NUMBER",
    REDUCER_SET_PROMOTION: "REDUCER_SET_PROMOTION",
    REDUCER_SET_CURRENT_QUESTION: "REDUCER_SET_CURRENT_QUESTION",
    REDUCER_SET_CURRENT_OPTIONS: "REDUCER_SET_CURRENT_OPTIONS",
    REDUCER_SET_TOKEN: "REDUCER_SET_TOKEN",
    REDUCER_SET_QUESTIONS_DATAS: "REDUCER_SET_QUESTIONS_DATAS",
    REDUCER_SET_QUIZ_ERROR: "REDUCER_SET_QUIZ_ERROR",

    REDUCER_RESET_QUIZ: "REDUCER_RESET_QUIZ",

    
    SAGA_GET_NEW_QUESTION: "SAGA_GET_NEW_QUESTION",
    SAGA_CHECK_ANSWER: "SAGA_CHECK_ANSWER",
    SAGA_UPDATE_QUIZ_REPORT: "SAGA_UPDATE_QUIZ_REPORT",
    SAGA_UPDATE_QUIZ_RESUME_REPORT: "SAGA_UPDATE_QUIZ_RESUME_REPORT",
    SAGA_GO_TO_NEXT_QUESTION: "SAGA_GO_TO_NEXT_QUESTION",
    SAGA_FINISH_QUIZ: "SAGA_FINISH_QUIZ",
    SAGA_RESUME_QUIZ: "SAGA_RESUME_QUIZ",
}
