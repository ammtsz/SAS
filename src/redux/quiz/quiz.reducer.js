import { QuizActionsTypes } from "./quiz.types";

const INITIAL_STATE = {
  active: false,
  loading: false,
  category: null,
  difficulty: "medium",
  questionNumber: 1,
  promotion: "1", //0 = got last one wrong, 1 = first question on that level, 2 = got last one right
  currentQuestion: {},
  currentOptions: [],
  questionsAnswered: [],
  error: null,
};

const quizReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case QuizActionsTypes.REDUCER_SET_ACTIVE:
      return { ...state, active: action.payload };

    case QuizActionsTypes.REDUCER_SET_LOADING:
      return { ...state, loading: action.payload };

    case QuizActionsTypes.REDUCER_SET_CATEGORY:
      return { ...state, category: action.payload };

    case QuizActionsTypes.REDUCER_SET_DIFFICULTY:
      return { ...state, difficulty: action.payload };

    case QuizActionsTypes.REDUCER_SET_QUESTION_NUMBER:
      return { ...state, questionNumber: action.payload };

    case QuizActionsTypes.REDUCER_SET_PROMOTION:
      return { ...state, promotion: action.payload };

    case QuizActionsTypes.REDUCER_SET_CURRENT_QUESTION:
      return { ...state, currentQuestion: action.payload };

    case QuizActionsTypes.REDUCER_SET_CURRENT_OPTIONS:
      return { ...state, currentOptions: action.payload };

    case QuizActionsTypes.REDUCER_SET_QUESTIONS_ANSWERED:
      return { ...state, questionsAnswered: action.payload };

    case QuizActionsTypes.REDUCER_RESET_QUIZ:
      return INITIAL_STATE;

    case QuizActionsTypes.REDUCER_SET_QUIZ_ERROR:
      return { ...state, error: action.payload };

    default:
      return state;
  }
};

export default quizReducer;
