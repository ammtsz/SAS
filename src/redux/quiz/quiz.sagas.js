import { all, call, takeLatest, put, select } from "redux-saga/effects";
import { fetchQuestion } from "../../api/trivia";
import { rsf } from "../../firebase/firebase.utils";
import {
  actionSetQuizError,
  actionSetQuizDifficulty,
  actionSetQuizPromotion,
  actionSetQuizCurrentQuestion,
  actionSetQuizCurrentOptions,
  actionSetQuizQuestionNumber,
  actionResetQuiz,
  actionSetQuizToken,
  actionSetQuizActive,
  actionSetQuizCategory,
  actionSetQuizQuestionsDatas,
  actionSetQuizLoading,
} from "./quiz.actions";
import { QuizActionsTypes } from "./quiz.types";
import {
  selectQuizCurrentQuestion,
  selectQuizDifficulty,
  selectQuizPromotion,
  selectQuizQuestionNumber,
  selectQuizCategory,
  selectQuizQuestionsDatas,
  selectQuizToken,
} from "./quiz.selectors";
import { selectReports } from "../categories/categories.selectors";
import { actionSetCategoriesReport } from "../categories/categories.actions";
import { selectUserDatas } from "../user/user.selectors";

// UTILS

export function* fetchQuestionFn(difficulty, category) {
  try {
    const token = yield select(selectQuizToken);
    const fetchedQuestion = yield fetchQuestion(difficulty, category, token);
    yield updateToken(token, fetchedQuestion.token);
    return fetchedQuestion.results;
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {};
  }
}
export function* updateToken(stateToken, fetchedToken) {
  try {
    if (stateToken !== fetchedToken) {
      yield put(actionSetQuizToken(fetchedToken));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* saveQuestionsDatas(fetchedQuestion) {
  try {
    yield put(actionSetQuizCurrentQuestion(fetchedQuestion));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* getCurrentOptions() {
  try {
    const currentQuestion = yield select(selectQuizCurrentQuestion);
    let allOptions = [];

    allOptions.push({ id: 0, option: currentQuestion.correct_answer });
    currentQuestion.incorrect_answers.forEach((option, index) => {
      allOptions.push({ id: index + 1, option });
    });

    yield put(
      actionSetQuizCurrentOptions(
        allOptions.sort((a, b) => (a.option > b.option ? 1 : -1))
      )
    );
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

export function* rightAnswerActions() {
  try {
    const quizPromotion = yield select(selectQuizPromotion);
    if (quizPromotion === "2") {
      yield increaseDifficulty();
      yield put(actionSetQuizPromotion("1"));
    } else {
      yield put(actionSetQuizPromotion("2"));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* increaseDifficulty() {
  try {
    const quizDifficulty = yield select(selectQuizDifficulty);
    switch (quizDifficulty) {
      case "easy":
        yield put(actionSetQuizDifficulty("medium"));
        break;
      case "medium":
        yield put(actionSetQuizDifficulty("hard"));
        break;
      default:
        yield put(actionSetQuizDifficulty(quizDifficulty));
        break;
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* wrongAnswerActions() {
  try {
    const quizPromotion = yield select(selectQuizPromotion);
    if (quizPromotion === "0") {
      yield decreaseDifficulty();
      yield put(actionSetQuizPromotion("1"));
    } else {
      yield put(actionSetQuizPromotion("0"));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* decreaseDifficulty() {
  const quizDifficulty = yield select(selectQuizDifficulty);
  try {
    switch (quizDifficulty) {
      case "hard":
        yield put(actionSetQuizDifficulty("medium"));
        break;
      case "medium":
        yield put(actionSetQuizDifficulty("easy"));
        break;
      default:
        yield put(actionSetQuizDifficulty(quizDifficulty));
        break;
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

export function* getReportDatas() {
  try {
    let easy = { rights: 0, wrongs: 0 };
    let medium = { rights: 0, wrongs: 0 };
    let hard = { rights: 0, wrongs: 0 };
    let rights = 0;
    let wrongs = 0;

    const quizQuestionsDatas = yield select(selectQuizQuestionsDatas);

    quizQuestionsDatas.forEach((question) => {
      if (question.difficulty === "easy") {
        if (question.selected_answer === "0") {
          easy = { ...easy, rights: easy.rights + 1 };
          rights += 1;
        } else {
          easy = { ...easy, wrongs: easy.wrongs + 1 };
          wrongs += 1;
        }
      } else if (question.difficulty === "medium") {
        if (question.selected_answer === "0") {
          medium = { ...medium, rights: medium.rights + 1 };
          rights += 1;
        } else {
          medium = { ...medium, wrongs: medium.wrongs + 1 };
          wrongs += 1;
        }
      } else if (question.difficulty === "hard") {
        if (question.selected_answer === "0") {
          hard = { ...hard, rights: hard.rights + 1 };
          rights += 1;
        } else {
          hard = { ...hard, wrongs: hard.wrongs + 1 };
          wrongs += 1;
        }
      }
    });

    return {
      difficulty: {
        easy: { rights: easy.rights, wrongs: easy.wrongs },
        medium: { rights: medium.rights, wrongs: medium.wrongs },
        hard: { rights: hard.rights, wrongs: hard.wrongs },
      },
      rights,
      wrongs,
    };
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {};
  }
}
export function* updateReportOnDB(report) {
  try {
    const userDatasState = yield select(selectUserDatas);
    if (userDatasState) {
      const userDatas = yield getUserDatasFromFirebase(userDatasState.id)
      yield call(rsf.firestore.updateDocument, `users/${userDatasState.id}`, {
        ...userDatas,
        reports: { ...userDatas.reports, ...report },
      });
    } else {
      localStorage.setItem("trivia", JSON.stringify(report));
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* getUserDatasFromFirebase(id) {
  try {
    const userSnapshot = yield call(
      rsf.firestore.getDocument,
      `users/${id}`
    );
    return userSnapshot.data();
  } catch (error) {
    yield put(actionSetQuizError(error));
    return {}
  }
}

// CALLED
export function* getNewQuestion({ payload: { difficulty, category } }) {
  try {
    yield put(actionSetQuizLoading(true));
    const fetchedQuestion = yield fetchQuestionFn(difficulty, category);
    yield saveQuestionsDatas(fetchedQuestion);
    yield getCurrentOptions(fetchedQuestion);
    yield put(actionSetQuizLoading(false));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* checkAnswer(selectedOption) {
  try {
    if (selectedOption.payload === "0") {
      yield rightAnswerActions();
    } else {
      yield wrongAnswerActions();
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* updateQuizReport() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    const reports = yield select(selectReports);
    const questions_datas = yield select(selectQuizQuestionsDatas);
    const token = yield select(selectQuizToken);
    const promotion = yield select(selectQuizPromotion);
    const results = yield getReportDatas();

    const resume =
      quizQuestionNumber < 10
        ? {
            token,
            promotion,
            difficulty,
          }
        : null;

    const categoryReport = {
      category_name: category.name,
      questions_datas,
      results,
      resume,
    };
    const newReport = { ...reports, [category.id]: categoryReport };
    yield put(actionSetCategoriesReport(newReport));
    yield updateReportOnDB(newReport);
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* updateQuizResumeReport() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    const reports = yield select(selectReports);
    const token = yield select(selectQuizToken);
    const promotion = yield select(selectQuizPromotion);
    const lastQuestion = yield select(selectQuizCurrentQuestion);

    const resume =
      quizQuestionNumber < 10
        ? {
            token,
            promotion,
            difficulty,
            lastQuestion,
          }
        : null;

    const newReport = {
      ...reports,
      [category.id]: { ...reports[category.id], resume },
    };
    yield put(actionSetCategoriesReport(newReport));
    yield updateReportOnDB(newReport);
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* goToNextQuestion() {
  try {
    const quizQuestionNumber = yield select(selectQuizQuestionNumber);
    const difficulty = yield select(selectQuizDifficulty);
    const category = yield select(selectQuizCategory);
    yield getNewQuestion({ payload: { difficulty, category: category.id } });
    yield updateQuizResumeReport();
    yield put(actionSetQuizQuestionNumber(quizQuestionNumber + 1));
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* finishQuiz() {
  try {
    yield put(actionResetQuiz());
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}
export function* resumeQuiz(category) {
  try {
    const reports = yield select(selectReports);
    const quizDatas = reports[category.payload.id];

    yield put(actionSetQuizActive(true));
    yield put(actionSetQuizCategory(category.payload));
    yield put(actionSetQuizDifficulty(quizDatas.resume.difficulty));
    yield put(actionSetQuizPromotion(quizDatas.resume.promotion));
    yield put(
      actionSetQuizQuestionNumber(quizDatas.questions_datas.length + 1)
    );
    yield put(actionSetQuizToken(quizDatas.resume.token));
    yield put(actionSetQuizQuestionsDatas(quizDatas.questions_datas));

    if (quizDatas.resume.lastQuestion) {
      yield put(actionSetQuizCurrentQuestion(quizDatas.resume.lastQuestion));
      yield getCurrentOptions();
    } else {
      yield getNewQuestion({
        payload: {
          difficulty: quizDatas.resume.difficulty,
          category: category.payload.id,
        },
      });
    }
  } catch (error) {
    yield put(actionSetQuizError(error));
  }
}

// CALLS
export function* onGetNewQuestion() {
  yield takeLatest(QuizActionsTypes.SAGA_GET_NEW_QUESTION, getNewQuestion);
}
export function* onCheckAnswer() {
  yield takeLatest(QuizActionsTypes.SAGA_CHECK_ANSWER, checkAnswer);
}
export function* onUpdateQuizReport() {
  yield takeLatest(QuizActionsTypes.SAGA_UPDATE_QUIZ_REPORT, updateQuizReport);
}
export function* onUpdateQuizResumeReport() {
  yield takeLatest(
    QuizActionsTypes.SAGA_UPDATE_QUIZ_RESUME_REPORT,
    updateQuizResumeReport
  );
}
export function* onGoToNextQuestion() {
  yield takeLatest(QuizActionsTypes.SAGA_GO_TO_NEXT_QUESTION, goToNextQuestion);
}
export function* onFinishQuiz() {
  yield takeLatest(QuizActionsTypes.SAGA_FINISH_QUIZ, finishQuiz);
}
export function* onResumeQuiz() {
  yield takeLatest(QuizActionsTypes.SAGA_RESUME_QUIZ, resumeQuiz);
}

export function* quizSagas() {
  yield all([
    call(onGetNewQuestion),
    call(onCheckAnswer),
    call(onUpdateQuizReport),
    call(onUpdateQuizResumeReport),
    call(onGoToNextQuestion),
    call(onFinishQuiz),
    call(onResumeQuiz),
  ]);
}
