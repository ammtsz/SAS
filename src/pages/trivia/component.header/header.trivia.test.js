import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import HeaderTrivia from "./header.trivia.component";

import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";

import { actionFinishQuiz } from "../../../redux/quiz/quiz.actions";

describe("Component trivia-page <HeaderTrivia />", () => {
  const mockStore = configureMockStore();
  const mockCategory = { id: 1, name: "Animals", completed: 4 }

  const store = mockStore({
    quiz: {
      category: mockCategory,
    },
  });
  store.dispatch = jest.fn()

  let component
  beforeEach(() => {
    component = render(
      <Provider store={store}>
        <HeaderTrivia />
      </Provider>
    );
  });

  it("should render HeaderTrivia", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should contain category name", () => {
    expect(screen.getByText(mockCategory.name)).toBeInTheDocument();
  });
  
  it("should contain close button", () => {
    expect(screen.getByText("Close")).toBeInTheDocument();
    expect(screen.getByAltText("close")).toBeInTheDocument();
  });
  
  it("should fire 'finishQuiz()' after clicking on close button", () => {
    const button = screen.getByTestId("trivia-close-button");
    fireEvent.click(button);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(actionFinishQuiz());
  });
});
