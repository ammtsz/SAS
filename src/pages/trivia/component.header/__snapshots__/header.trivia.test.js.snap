// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component trivia-page <HeaderTrivia /> should render HeaderTrivia 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ fvHmtS"
  >
    <h4
      class="sc-gsTCUz helITq"
    >
      Animals
    </h4>
    <button
      class="sc-dlfnbm iWBeHa"
      data-testid="trivia-close-button"
    >
      <img
        alt="close"
        class="sc-hKgILt jjqRdP"
        src="close-btn.svg"
        tabindex="-1"
      />
      Close
    </button>
  </div>
</DocumentFragment>
`;
